server/src/main/java/org/elasticsearch/index/query/DistanceFeatureQueryBuilder.java

```diff
@@ -0,0 +1,190 @@
+import org.elasticsearch.index.mapper.DateFieldMapper.DateFieldType;
+import org.elasticsearch.index.mapper.NumberFieldMapper.NumberFieldType;
+import org.elasticsearch.index.mapper.NumberFieldMapper.NumberType;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
+/**
+ * A query to boost scores based on their proximity to the given origin
+ */
+public class DistanceFeatureQueryBuilder extends AbstractQueryBuilder<DistanceFeatureQueryBuilder> {
+    public static final String NAME = "distance_feature";
+
+    private static final ParseField FIELD_FIELD = new ParseField("field");
+    private static final ParseField ORIGIN_FIELD = new ParseField("origin");
+    private static final ParseField PIVOT_FIELD = new ParseField("pivot");
+
+    private final String field;
+    private final Object origin; //type of Long, String or GeoPoint
+    private final Object pivot; //type of Long or String
+
+    private static final ConstructingObjectParser<DistanceFeatureQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "distance_feature", false,
+        args -> new DistanceFeatureQueryBuilder((String) args[0], args[1], args[2])
+    );
+
+    static {
+        PARSER.declareString(constructorArg(), FIELD_FIELD);
+        // origin: number for long fields; number or string for date fields; string, array, object for geo fields
+        PARSER.declareField(constructorArg(), DistanceFeatureQueryBuilder::originFromXContent,
+            ORIGIN_FIELD, ObjectParser.ValueType.OBJECT_ARRAY_STRING_OR_NUMBER);
+        //pivot: number for long fields; string for dates or geo fields
+        PARSER.declareField(constructorArg(), DistanceFeatureQueryBuilder::pivotFromXContent,
+            PIVOT_FIELD, ObjectParser.ValueType.LONG);
+        declareStandardFields(PARSER);
+    }
+
+    public DistanceFeatureQueryBuilder(String field, Object origin, Object pivot) {
+        this.field = field;
+        this.origin = origin;
+        this.pivot = pivot;
+    }
+
+    public static DistanceFeatureQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            DistanceFeatureQueryBuilder builder = PARSER.apply(parser, null);
+            return builder;
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    public static Object originFromXContent(XContentParser parser) throws IOException {
+        if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            return parser.longValue();
+        } else if(parser.currentToken() == XContentParser.Token.VALUE_STRING) {
+            return parser.text();
+        } else if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            return GeoUtils.parseGeoPoint(parser);
+        } else { // if currentToken equals to START_ARRAY)
```

Comment:
Can you check START_ARRAY explicitely here and otherwise throw an error if we encounter an unexpected token?